src/codegen/transformations.lisp

```diff
@@ -7,208 +7,200 @@
+    (action (:traverse node-let node traverse &rest args)
+      (make-node-let
+       :type (node-type node)
+       :bindings (loop :for (name . node) :in (node-let-bindings node)
+                       :collect (cons name (apply traverse node args)))
+       :subexpr (apply traverse (node-let-subexpr node) args)))
+    (action (:traverse node-match node traverse &rest args)
+      (make-node-match
+       :type (node-type node)
+       :expr (apply traverse (node-match-expr node) args)
+       :branches (mapcar
+                  (lambda (branch)
+                    (make-match-branch
+                     :pattern (match-branch-pattern branch)
+                     :body (apply traverse
+                                  (match-branch-body branch)
+                                  args)))
+                  (node-match-branches node))))
+    (action (:traverse node-while node traverse &rest args)
+      (make-node-while
+       :type (node-type node)
+       :label (node-while-label node)
+       :expr (apply traverse (node-while-expr node) args)
+       :body (apply traverse (node-while-body node) args)))
+    (action (:traverse node-while-let node traverse &rest args)
+      (make-node-while-let
+       :type (node-type node)
+       :label (node-while-let-label node)
+       :pattern (node-while-let-pattern node)
+       :expr (apply traverse (node-while-let-expr node) args)
+       :body (apply traverse (node-while-let-body node) args)))
+    (action (:traverse node-loop node traverse &rest args)
+      (make-node-loop
+       :type (node-type node)
+       :label (node-loop-label node)
+       :body (apply traverse (node-loop-body node) args)))
+    (action (:traverse node-seq node traverse &rest args)
+      (make-node-seq
+       :type (node-type node)
+       :nodes (mapcar
+               (lambda (node)
+                 (apply traverse node args))
+               (node-seq-nodes node))))
+    (action (:traverse node-return node traverse &rest args)
+      (make-node-return
+       :type (node-type node)
+       :expr (apply traverse (node-return-expr node) args)))
+    (action (:traverse node-field node traverse &rest args)
+      (make-node-field
+       :type (node-type node)
+       :name (node-field-name node)
+       :dict (apply traverse (node-field-dict node) args)))
+    (action (:traverse node-dynamic-extent node traverse &rest args)
+      (make-node-dynamic-extent
+       :type (node-type node)
+       :name (node-dynamic-extent-name node)
+       :node (apply traverse (node-dynamic-extent-node node) args)
+       :body (apply traverse (node-dynamic-extent-body node) args)))
+    (action (:traverse node-bind node traverse &rest args)
+      (make-node-bind
+       :type (node-type node)
+       :name (node-bind-name node)
+       :expr (apply traverse (node-bind-expr node) args)
+       :body (apply traverse (node-bind-body node) args))))
+   t))
+
+(defun optional-call (when-key type-key actions traverse args node)
```

Comment:
maybe now we can use the "action" terminology now? `fire-actions` `fire-matching-actions` maybe something better